### DISCUSSION TITLE
Disable DHCP option 60

### DISCUSSION THREAD

**Discussion by igorsvetlana** [ipxe/ipxe#1279] [Q&A] [answered]
Hello everyone,

I know this is not the right forum but ill give it a shot anyways.

We are currently migrating from VMware to XenServer and we encountered a problem when PXE booting hosts from XenServer to Citrix PVS. In VMWare it works but not with XenServer, we found that the hosts from XenServer "asks" for the DHCP option 66 and VMware does not. If we configure option 66 in the DHCP server it works, but this is not what we are after.

Citrix themselves says that it's not possible to disable option 60 when its booting from IPXE, but I can't see why, isn't it possible to modify the "configuration" for IPXE?

Thanks in advance

**Comment by mcb30** [comment, marked as answer]
You haven't given any indication of why requesting option 66 is causing you any kind of problem.  If the DHCP server does not have any value configured for option 66, then it will just not return a value for that option.  It will not cause any problem or prevent the boot from completing.

**Reply by igorsvetlana** [reply]
We need to use DHCP option 60 for HA, if we use option 66 we can only specify one of the PVS servers (we can go around this by load balancing the PVS servers). 

In our testing the only difference between VMware and Xen is that Xen "asks" for option 66 and if we don't have this value configured on the DHCP server it wont boot (it says no file found and points us to https://ipxe.org/err/3c0920) as soon as we add option 66 it boots, so I just thought that we'll just disable the DHCP option 66 request and it will work? 


![2024-08-16 12_36_00-Window](https://github.com/user-attachments/assets/c681d91c-9c1a-45a3-93e4-1128cda92f49)
![2024-08-16 12_35_21-Window](https://github.com/user-attachments/assets/d78127fd-2bfe-4fd9-899f-851f62c4634f)

**Reply by mcb30** [reply]
> We need to use DHCP option 60 for HA, if we use option 66 we can only specify one of the PVS servers (we can go around this by load balancing the PVS servers).

DHCP option 60 is the vendor class identifier.  The server-provided option should be the fixed value `PXEClient` for any PXE boot.  I don't see how this fixed value can have any relevance to HA.

In general: your issue report provides your opinion of what the problem might be, but doesn't actually show the issue that you are experiencing.  Take a look at https://www.chiark.greenend.org.uk/~sgtatham/bugs.html and try giving a fresh description of your problem.  In particular: what do you actually **see** in the error case?  Send a screenshot of the full output from iPXE, not a screenshot from a small part of your wireshark capture.

**Comment by igorsvetlana** [comment]
My title is wrong, sorry... I want to disable option 66 and keep 60. We have multiple PVS servers with DHCP role installed on each, DHCP settings are synced across. If we use option 66 we can only specify 1 IP, if we use option 60 we can specify multiple IPs, but to the problem IPXE will not boot if we don't have option 66 specified. So I just thought that disabling option 66 in IPXE when it sends the discover packet will solve the issue.

**Comment by mcb30** [comment]
> My title is wrong, sorry... I want to disable option 66 and keep 60. We have multiple PVS servers with DHCP role installed on each, DHCP settings are synced across. If we use option 66 we can only specify 1 IP, if we use option 60 we can specify multiple IPs, but to the problem IPXE will not boot if we don't have option 66 specified. So I just thought that disabling option 66 in IPXE when it sends the discover packet will solve the issue.

DHCP option 60 is the vendor class identifier.  The server-provided option should be the fixed value `PXEClient` for any PXE boot.  I don't see how this fixed value can have any relevance to HA.

In general: your issue report provides your opinion of what the problem might be, but doesn't actually show the issue that you are experiencing.  Take a look at https://www.chiark.greenend.org.uk/~sgtatham/bugs.html and try giving a fresh description of your problem.  In particular: what do you actually **see** in the error case?  Send a screenshot of the full output from iPXE, not a screenshot from a small part of your wireshark capture.

**Reply by igorsvetlana** [reply]
I read the answer the first time, I don't know what more you want I only want to know if its possible to edit the dhcp boots options inside the ipxe.iso and if so how do you do it. 
This is what it looks like when booting without option 66 configured on DHCP server.
![2024-08-15 10_46_40-Window](https://github.com/user-attachments/assets/4accba2f-a2ca-41ad-992d-449d64793f59)

**Reply by mcb30** [reply]
> This is what it looks like when booting without option 66 configured on DHCP server. ![2024-08-15 10_46_40-Window](https://private-user-images.githubusercontent.com/23136475/360397976-4accba2f-a2ca-41ad-992d-449d64793f59.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzMjU2ODAsIm5iZiI6MTcyNDMyNTM4MCwicGF0aCI6Ii8yMzEzNjQ3NS8zNjAzOTc5NzYtNGFjY2JhMmYtYTJjYS00MWFkLTk5MmQtNDQ5ZDY0NzkzZjU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIyVDExMTYyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhODRhMmRiNGU4MTBmNDdkMmQ2N2VmOTI1NTNkZWQwMzdkZGMwYWFkYmI0ZDE4YmMzOTY4ZGYzZWY4YzZhOWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7IwQ55y9xHY0F_OP3PFhkloD3q_3rs9akxbnhJ1cfeU)

Thank you for providing the screenshot.  That is useful information.

The screenshot shows that you are providing a TFTP filename (`ARDBP32.BIN`) but you are not providing any TFTP server IP address.  In `dhcpd` terms, you have provided `filename` but you have not provided `next-server`.  This is why your boot is failing.

iPXE does not use the value of DHCP option 66 (the TFTP server name) in any way.  It is possible that your DHCP server performs an implicit DNS lookup when you set a value for option 66, and automatically sets the `next-server` field to contain the resolved IP address.

I would be interested to also see a screenshot from when you **do** provide option 66.

**Comment by igorsvetlana** [comment]
We don't have anything specified at all in option 60 (this works in VMware). I've also tried to add one server IP and multiple IPs in this field but same results. As soon as I populate the option 66 with one IP it boots as usual. And like I said I know we can load balance all IPs, but im curious why this works in VMware and not XenServer...

[https://support.citrix.com/s/article/CTX208519-configuring-pvs-for-high-availability-with-uefi-booting-and-pxe-service?language=en_US](url)
"PXE server (option 60, doesn’t need to be configured)
PXE server responds to DHCP request with PXE information, giving its own server name, and the appropriate file name for tftp retrieval of the pxe bootstrap file"

![2024-08-22 15_00_45-](https://github.com/user-attachments/assets/ca90ec06-1391-465b-939d-8d21fdaec891)
Boot with above DHCP settings
![2024-08-22 14_59_18-](https://github.com/user-attachments/assets/21097389-5a23-4f39-9c1b-8dfa0e5150c1)

**Reply by igorsvetlana** [reply]
isnt it just possible to remove these lines somehow? :|

\ipxe\src\net\dhcppkt.c
\ipxe\src\net\udp\dhcp.c
![2024-08-23 10_37_31](https://github.com/user-attachments/assets/dbd87985-cd7e-4164-ab98-765357fcbced)
![2024-08-23 10_36_04](https://github.com/user-attachments/assets/ef5bfb60-dea1-4bdf-b74d-1ee856e0dae5)

**Reply by mcb30** [reply]
> We don't have anything specified at all in option 60 (this works in VMware).

Please drop this misguided focus on DHCP option 60.  As I have already told you, option 60 is the DHCP vendor class, and has nothing to do with the problem you are experiencing.

> ![2024-08-22 15_00_45-](https://private-user-images.githubusercontent.com/23136475/360439579-ca90ec06-1391-465b-939d-8d21fdaec891.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0MDI4NTksIm5iZiI6MTcyNDQwMjU1OSwicGF0aCI6Ii8yMzEzNjQ3NS8zNjA0Mzk1NzktY2E5MGVjMDYtMTM5MS00NjViLTkzOWQtOGQyMWZkYWVjODkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDA4NDIzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgyODI5ZGU1ZmYzNGNiOWJiMWIxMTQ0MzQ2NjNhN2U5ZTBlMDU3ODZmNTJhY2U2YWExODJjYzNhNmM2YTFiMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WFAFzu--rpS1RkVt_P14UTCCsMxctyS6Jl4OKNmYRm4)

OK, you are using some version of the Microsoft DHCP server.  From memory, Microsoft doesn't provide a specific way to configure the `next-server` address, but will use the value that you configure for DHCP option 66 to populate the `next-server` field.

So: when you actually configure the TFTP server IP address (which is done by configuring a value for DHCP option 66 in your choice of DHCP server software), you are able to boot.  When you don't configure a TFTP server address, you are not able to boot.  This is correct behaviour.  The defective behaviour is in whatever other software you are using that somehow manages to boot from a TFTP server without being told which TFTP server to use.

But you don't seem to be interested in learning about the real solution to your problem, and I've run out of patience, sorry.  I've given you the answers you need are above, if you care enough to read them.

**Reply by igorsvetlana** [reply]
I'm sorry, i read everything again i mixed things up. I thought that the Citrix article said that if i dont populate a value in option 60 it will give its own server name in option 60... I'm still trying to understand how this works in VMwares pxe boot and not with ipxe, but ill try to look in to. Thank you for the help
